executor/executor_test.go

```diff
@@ -8999,3 +8999,20 @@ func (s *testSuite) TestIssue25447(c *C) {
 	tk.MustExec("insert into t2(a) values(1)")
 	tk.MustQuery("select /*+ tidb_inlj(t2) */ t2.b, t1.b from t1 join t2 ON t2.a=t1.a").Check(testkit.Rows("<nil> 1"))
 }
+
+func (s *testSuite) TestIssue23602(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("USE test")
+	tk.Se.GetSessionVars().EnableClusteredIndex = variable.ClusteredIndexDefModeOn
+	tk.MustExec("CREATE TABLE t (a bigint unsigned PRIMARY KEY)")
+	defer tk.MustExec("DROP TABLE t")
+	tk.MustExec("INSERT INTO t VALUES (0),(1),(2),(3),(18446744073709551600),(18446744073709551605),(18446744073709551610),(18446744073709551615)")
+	tk.MustExec("ANALYZE TABLE t")
+	tk.MustQuery(`EXPLAIN FORMAT = 'brief' SELECT a FROM t WHERE a >= 0x1 AND a <= 0x2`).Check(testkit.Rows(
+		"TableReader 2.00 root  data:TableRangeScan]\n" +
+			"[└─TableRangeScan 2.00 cop[tikv] table:t range:[1,2], keep order:false"))
+	tk.MustQuery(`EXPLAIN FORMAT = 'brief' SELECT a FROM t WHERE a BETWEEN 0x1 AND 0x2`).Check(testkit.Rows(
+		"TableReader 2.00 root  data:TableRangeScan]\n" +
+			"[└─TableRangeScan 2.00 cop[tikv] table:t range:[1,2], keep order:false"))
+	tk.MustQuery("SELECT a FROM t WHERE a BETWEEN 0xFFFFFFFFFFFFFFF5 AND X'FFFFFFFFFFFFFFFA'").Check(testkit.Rows("18446744073709551605", "18446744073709551610"))
```

Comment:
drop table if exists t1;
create table t1(c1 bigint unsigned);                                                                                                                                                                                                                                          insert into t1 values(18446744073709551610);
select * from t1 where c1 between 18446744073709551601 and 0xFFFFFFFFFFFFFFFE;

Looks like not compatible with MySQL 8.0.

Reply:
That seems to be a bug in MySQL?
```sql
mysql> drop table if exists t1;
Query OK, 0 rows affected (0,06 sec)

mysql> create table t1(c1 bigint unsigned); insert into t1 values(18446744073709551610);
Query OK, 0 rows affected (0,06 sec)

Query OK, 1 row affected (0,02 sec)

mysql> select * from t1 where c1 between 18446744073709551601 and 0xFFFFFFFFFFFFFFFE;
Empty set (0,00 sec)

mysql> explain select * from t1 where c1 between 18446744073709551601 and 0xFFFFFFFFFFFFFFFE;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | t1    | NULL       | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using where |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0,00 sec)

mysql> show warnings;
+-------+------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| Level | Code | Message                                                                                                                                              |
+-------+------+------------------------------------------------------------------------------------------------------------------------------------------------------+
| Note  | 1003 | /* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` between 18446744073709551601 and 18446744073709551614) |
+-------+------+------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0,00 sec)

mysql> select version();
+-------------------------+
| version()               |
+-------------------------+
| 8.0.26-0ubuntu0.21.04.3 |
+-------------------------+
1 row in set (0,00 sec)

mysql> /* select#1 */ select `test`.`t1`.`c1` AS `c1` from `test`.`t1` where (`test`.`t1`.`c1` between 18446744073709551601 and 18446744073709551614);
+----------------------+
| c1                   |
+----------------------+
| 18446744073709551610 |
+----------------------+
1 row in set (0,01 sec)
```

I consider TiDB to do the right thing after this patch, and suggest to not be bug compatible with MySQL for this case.
@guo-shaoge what do you think?